File: external/boost_1.55.0/libs/python/test/auto_ptr.py
# Copyright David Abrahams 2004. Distributed under the Boost
# Software License, Version 1.0. (See accompanying
# file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
'''
>>> from auto_ptr_ext import *
>>> x = X(42)
>>> x.value()
42
>>> look(x), look(x)
(42, 42)

>>> maybe_steal(x, 0)
42
>>> look(x)
42

>>> maybe_steal(x, 1)
42
>>> broken_auto_ptr and -1 or look(x)
-1

>>> x = X(69)
>>> steal(x)
69
>>> broken_auto_ptr and -1 or look(x)
-1

>>> if not broken_auto_ptr:
...     try: x.value()
...     except TypeError: pass
...     else: print 'expected a TypeError exception'

>>> x = make()
>>> look(x)
77

>>> z = callback(lambda z: z)
>>> z.value()
77

>>> extract(x).value()
77

#
# Test derived to base conversions
#

>>> y = Y(42)
>>> y.value()
42

>>> try: maybe_steal(y, 0)
... except TypeError: pass
... else: print 'expected a TypeError exception'

>>> y.value()
42

>>> broken_auto_ptr and 42 or steal(y)
42

>>> if not broken_auto_ptr:
...     try: y.value()
...     except TypeError: pass
...     else: print 'expected a TypeError exception'

>>> print look.__doc__.splitlines()[1]
look( (X)arg1) -> int :

>>> print steal.__doc__.splitlines()[1]
steal( (X)arg1) -> int :

>>> print maybe_steal.__doc__.splitlines()[1]
maybe_steal( (X)arg1, (bool)arg2) -> int :

>>> print make.__doc__.splitlines()[1]
make() -> X :

>>> print callback.__doc__.splitlines()[1]
callback( (object)arg1) -> X :

>>> print extract.__doc__.splitlines()[1]
extract( (object)arg1) -> X :

'''

def run(args = None):
    import sys
    import doctest

    if args is not None:
        sys.argv = args
    return doctest.testmod(sys.modules.get(__name__))
    
if __name__ == '__main__':
    print "running..."
    import sys
    status = run()[0]
    if (status == 0): print "Done."
    sys.exit(status)

File: external/boost_1.55.0/libs/python/test/auto_ptr.cpp
// Copyright David Abrahams 2002.
// Distributed under the Boost Software License, Version 1.0. (See
// accompanying file LICENSE_1_0.txt or copy at
// http://www.boost.org/LICENSE_1_0.txt)

#include <boost/python/module.hpp>
#include "test_class.hpp"
#include <boost/python/class.hpp>
#include <boost/python/extract.hpp>
#include <boost/python/def.hpp>
#include <boost/python/implicit.hpp>

#include <boost/detail/workaround.hpp>

#include <memory>

using namespace boost::python;

typedef test_class<> X;

struct Y : X
{
    Y(int n) : X(n) {};
};

int look(std::auto_ptr<X> const& x)
{
    return (x.get()) ? x->value() : -1;
}

int steal(std::auto_ptr<X> x)
{
    return x->value();
}

int maybe_steal(std::auto_ptr<X>& x, bool doit)
{
    int n = x->value();
    if (doit)
        x.release();
    return n;
}

std::auto_ptr<X> make()
{
    return std::auto_ptr<X>(new X(77));
}

std::auto_ptr<X> callback(object f)
{
    std::auto_ptr<X> x(new X(77));
    return call<std::auto_ptr<X> >(f.ptr(), x);
}

std::auto_ptr<X> extract_(object o)
{
    return extract<std::auto_ptr<X>&>(o)
#if BOOST_MSVC <= 1300
        ()
#endif 
        ;
}

BOOST_PYTHON_MODULE(auto_ptr_ext)
{
    class_<X, std::auto_ptr<X>, boost::noncopyable>("X", init<int>())
        .def("value", &X::value)
        ;
    
    class_<Y, std::auto_ptr<Y>, bases<X>, boost::noncopyable>("Y", init<int>())
        ;

    // VC6 auto_ptrs do not have converting constructors    
#if BOOST_WORKAROUND(BOOST_DINKUMWARE_STDLIB, < 306)
    scope().attr("broken_auto_ptr") = 1;
#else
    scope().attr("broken_auto_ptr") = 0;
    implicitly_convertible<std::auto_ptr<Y>, std::auto_ptr<X> >();
#endif
    
    def("look", look);
    def("steal", steal);
    def("maybe_steal", maybe_steal);
    def("make", make);
    def("callback", callback);
    def("extract", extract_);
}

#include "module_tail.cpp"

